contracts/terraswap_router/src/contract.rs

```diff
@@ -317,6 +330,154 @@ fn simulate_swap_operations(
+            SwapOperation::Loop {
+                offer_asset_info,
+                ask_asset_info,
+            } => {
+                let loop_factory = deps.api.addr_humanize(&config.loop_factory)?;
+
+                reverse_simulate_return_amount(
+                    deps,
+                    loop_factory,
+                    ask_amount,
+                    offer_asset_info,
+                    ask_asset_info,
+                )
+                .unwrap()
+            }
+            SwapOperation::Astroport {
+                offer_asset_info,
+                ask_asset_info,
+            } => {
+                let astroport_factory = deps.api.addr_humanize(&config.astroport_factory)?;
+
+                reverse_simulate_return_amount(
+                    deps,
+                    astroport_factory,
+                    ask_amount,
+                    offer_asset_info,
+                    ask_asset_info,
+                )
+                .unwrap()
+            }
+        }
+    }
+
+    Ok(SimulateSwapOperationsResponse { amount: ask_amount })
+}
+
+fn simulate_return_amount(
+    deps: Deps,
+    factory: Addr,
+    mut offer_amount: Uint128,
+    offer_asset_info: AssetInfo,
+    ask_asset_info: AssetInfo,
+) -> StdResult<Uint128> {
+    let pair_info: PairInfo = query_pair_info(
+        &deps.querier,
+        factory,
+        &[offer_asset_info.clone(), ask_asset_info.clone()],
+    )?;
+
+    // Deduct tax before querying simulation
+    if let AssetInfo::NativeToken { denom } = offer_asset_info.clone() {
+        offer_amount =
+            offer_amount.checked_sub(compute_tax(&deps.querier, offer_amount, denom)?)?;
+    }
+
+    let mut res: SimulationResponse =
+        deps.querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
+            contract_addr: pair_info.contract_addr,
+            msg: to_binary(&PairQueryMsg::Simulation {
+                offer_asset: Asset {
+                    info: offer_asset_info,
+                    amount: offer_amount,
+                },
+            })?,
+        }))?;
+
+    // Deduct tax after querying simulation
+    if let AssetInfo::NativeToken { denom } = ask_asset_info {
+        res.return_amount =
+            res.return_amount
+                .checked_sub(compute_tax(&deps.querier, res.return_amount, denom)?)?;
+    }
+
+    Ok(res.return_amount)
+}
+
+fn reverse_simulate_return_amount(
+    deps: Deps,
+    factory: Addr,
+    ask_amount: Uint128,
+    offer_asset_info: AssetInfo,
+    ask_asset_info: AssetInfo,
+) -> StdResult<Uint128> {
+    let pair_info: PairInfo = query_pair_info(
+        &deps.querier,
+        factory,
+        &[offer_asset_info.clone(), ask_asset_info.clone()],
+    )?;
+
+    let mut res = reverse_simulate(
+        &deps.querier,
+        Addr::unchecked(pair_info.contract_addr),
+        &Asset {
+            amount: ask_amount,
+            info: ask_asset_info,
+        },
+    )?;
+
+    // Deduct tax after querying simulation
```

Comment:
needs update comment 'deduct' to 'add'?

Reply:
https://github.com/terraswap/terraswap/pull/34/commits/7d56451c033fb07288eda67115e31a6742c76e00